nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/OAuth2Object.java

```diff
@@ -0,0 +1,165 @@
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Endpoint to retrieve OAuth2 authentication data
+ * @since 8.4
+ */
+@WebObject(type = "oauth2")
+public class OAuth2Object extends AbstractResource<ResourceTypeImpl> {
+
+    /**
+     * Retrieves oauth2 data for a given provider.
+     */
+    @GET
+    @Path("provider/{providerId}")
+    public Response getProvider(@PathParam("providerId") String providerId,
+                                @Context HttpServletRequest request) throws IOException, RestOperationException {
+
+        NuxeoOAuth2ServiceProvider provider = getProvider(providerId);
+
+        Map<String,Object> result = new HashMap<>();
+        result.put("serviceName", provider.getServiceName());
+        result.put("enabled", provider.isEnabled());
+        result.put("clientId", provider.getClientId());
+        result.put("scopes", provider.getScopes());
+        result.put("authorizationURL", provider.getClientId() == null ? null : provider.getAuthorizationUrl(request));
+        result.put("authorizationServerURL", provider.getAuthorizationServerURL());
+        result.put("tokenServerURL", provider.getTokenServerURL());
+
+        String username = request.getUserPrincipal().getName();
+        NuxeoOAuth2Token token = getToken(provider, username);
+        boolean hasToken = (token != null);
+        String login = hasToken ? token.getServiceLogin() : null;
+        result.put("hasToken", hasToken);
+        result.put("userId", login);
+
+        return buildResponse(Status.OK, result);
+    }
+
+    /**
+     * Retrieves a valid access token for a given provider and the current user.
+     * If expired, the token will be refreshed.
+     */
+    @GET
+    @Path("provider/{providerId}/token")
+    public Response getToken(@PathParam("providerId") String providerId,
+                             @Context HttpServletRequest request) throws IOException, RestOperationException {
+
+        NuxeoOAuth2ServiceProvider provider = getProvider(providerId);
+
+        String username = request.getUserPrincipal().getName();
+        NuxeoOAuth2Token token = getToken(provider, username);
+        if (token == null) {
+            return Response.status(Status.NOT_FOUND).build();
+        }
+        Credential credential = getCredential(provider, token);
+
+        if (credential == null) {
+            return Response.status(Status.NOT_FOUND).build();
+        }
+        Long expiresInSeconds = credential.getExpiresInSeconds();
+        if (expiresInSeconds != null && expiresInSeconds <= 0) {
+            credential.refreshToken();
+        }
+        Map<String,Object> result = new HashMap<>();
+        result.put("token", credential.getAccessToken());
+        return buildResponse(Status.OK, result);
+
+    }
+
+    private NuxeoOAuth2Token getToken(NuxeoOAuth2ServiceProvider provider, String nxuser) {
```

Comment:
The getServiceUserId method can either worth with service or nuxeo users, according to whether it extends AbstractOAuth2UserEmailProvider or not. Because the nuxeo user is the only thing we have when we receive the request and we might be working with an AbstractOAuth2UserEmailProvider, we have to perform a query on the credential data store to get the token for the given nuxeo user and provider.